modules/fedimint-lnv2-client/src/lib.rs

```diff
@@ -0,0 +1,725 @@
+        reqwest::Client::new()
+            .post(
+                gateway_api
+                    .join("payment_info")
+                    .expect("'payment_info' contains no invalid characters for a URL")
+                    .as_str(),
+            )
+            .json(&self.federation_id)
+            .send()
+            .await?
+            .json::<Option<PaymentInfo>>()
+            .await?
+            .ok_or(anyhow!("The gateway does not support our federation"))
+    }
+
+    pub async fn send(
+        &self,
+        gateway_api: SafeUrl,
+        invoice: Bolt11Invoice,
+    ) -> anyhow::Result<OperationId> {
+        self.send_internal(
+            gateway_api,
+            invoice,
+            PaymentFee::default(),
+            SEND_EXPIRATION_DELTA_BLOCKS_DEFAULT,
+        )
+        .await
+    }
+
+    pub async fn send_internal(
+        &self,
+        gateway_api: SafeUrl,
+        invoice: Bolt11Invoice,
+        max_fee: PaymentFee,
+        expiration_delta: u64,
+    ) -> anyhow::Result<OperationId> {
+        let (ephemeral_tweak, ephemeral_pk) = generate_ephemeral_tweak(self.keypair.public_key());
+
+        let refund_keypair = SecretKey::from_slice(&ephemeral_tweak)
+            .expect("32 bytes, within curve order")
+            .keypair(secp256k1::SECP256K1);
+
+        let payment_info = self
+            .payment_info(gateway_api.clone())
+            .await
+            .context("Gateway is offline")?;
+
+        ensure!(
+            max_fee >= payment_info.payment_fees.send,
+            "The gateway send fee is above {max_fee:?}"
+        );
+
+        // we double it to account for cltv expiry deltas along the lightning route
+        let min_outgoing_cltv_delta = 2 * payment_info.outgoing_cltv_delta;
+
+        ensure!(
+            min_outgoing_cltv_delta <= expiration_delta,
+            "The minimum possible expiration delta is {min_outgoing_cltv_delta} blocks"
+        );
+
+        let consensus_block_count = self
+            .module_api
+            .consensus_block_count()
+            .await
+            .context("Failed to fetch the consensus block count from the federation")?;
+
+        let invoice_msats = invoice
+            .amount_milli_satoshis()
+            .ok_or(anyhow!("Invoice has no amount"))?;
+
+        let contract_amount = payment_info
+            .payment_fees
+            .outgoing_contract_amount(invoice_msats);
+
+        let contract = OutgoingContract {
+            payment_hash: *invoice.payment_hash(),
+            amount: contract_amount,
+            expiration: consensus_block_count + expiration_delta,
+            claim_pk: payment_info.public_key,
+            refund_pk: refund_keypair.public_key(),
+            ephemeral_pk,
+            invoice_hash: invoice.consensus_hash(),
+        };
+
+        let operation_id = OperationId(contract.consensus_hash::<sha256::Hash>().into_inner());
```

Comment:
In the legacy client we store a `PaymentResult` that keeps track of successfully paid invoices, so that we don't try to pay the same invoice twice. I think we will need something like that here too.